components/drivers/audio/audio.c

```diff
@@ -0,0 +1,843 @@
+/*
+ * audio.c
+ *
+ *  Created on: 2016��10��19��
+ *      Author: Urey
+ */
+
+#include <stdio.h>
+#include <rthw.h>
+#include <rtthread.h>
+#include <rtdevice.h>
+
+#include <drivers/audio.h>
+
+#define AUDIO_DEBUG   0
+#if AUDIO_DEBUG
+#define AUDIO_DBG(...)     printf("[AUDIO]:"),printf(__VA_ARGS__)
+#else
+#define AUDIO_DBG(...)
+#endif
+
+
+rt_err_t _audio_queue_init(struct rt_audio_queue *queue, rt_uint16_t size, rt_uint16_t lwm)
+{
+    RT_ASSERT(queue != RT_NULL);
+
+    queue->count 	= 0;
+    queue->size  	= size;
+    queue->lwm 		= lwm;
+    queue->waiting_lwm 	= RT_FALSE;
+
+    queue->get_index 	= 0;
+    queue->put_index 	= 0;
+
+    rt_list_init(&(queue->suspended_push_list));
+    rt_list_init(&(queue->suspended_pop_list));
+
+    queue->queue = (struct rt_audio_frame *)rt_malloc(sizeof(struct rt_audio_frame) * size);
+    if (queue->queue == RT_NULL)
+    {
+        return -RT_ENOMEM;
+    }
+
+    return RT_EOK;
+}
+
+rt_err_t _audio_queue_push(struct rt_audio_queue *queue, struct rt_audio_frame *frame, rt_int32_t timeout)
+{
+    rt_ubase_t  level;
+    rt_thread_t thread;
+    rt_err_t    result;
+
+    RT_ASSERT(queue != RT_NULL);
+
+    result = RT_EOK;
+    thread = rt_thread_self();
+
+    AUDIO_DBG("%s count = %d\n",__func__,queue->count);
+
+    level = rt_hw_interrupt_disable();
+    while(queue->count == queue->size)
+    {// audio queue is full
+        queue->waiting_lwm = RT_TRUE;
+
+        /* queue is full */
+        if (timeout == 0)
+        {
+            result = -RT_ETIMEOUT;
+            goto __exit;
+        }
+
+        /* current context checking */
+        RT_DEBUG_NOT_IN_INTERRUPT;
+
+        /* reset thread error number */
+        thread->error = RT_EOK;
+
+        /* suspend thread on the push list */
+        rt_thread_suspend(thread);
+        rt_list_insert_before(&(queue->suspended_push_list), &(thread->tlist));
```

Comment:
Is there any strong reason that implement your own queue here instead of using the `dataqueue`?